src/gluonts/model/deepvar_hierarchical/_network.py

```diff
@@ -0,0 +1,341 @@
+        # If seq_axis = None, during training the first axis is only `batch_size`,
+        # in which case `M` would be expanded 3 times; during prediction it would be expanded 2 times since the first
+        # axis is `batch_size x  num_parallel_samples`.
+        M_expanded = self.M
+        for i in range(len(samples.shape[num_iter_dims:-1])):
+            M_expanded = M_expanded.expand_dims(axis=0)
+
+        # If seq_axis = None broadcast M to (num_samples, batch_size, seq_len, m, m) during training
+        # and to (num_samples * batch_size, seq_len, m, m) during prediction
+        # Else broadcast to the appropriate remaining dimension
+        _shape = (
+            list(samples.shape[:-1])
+            if not self.seq_axis
+            else [
+                samples.shape[i]
+                for i in range(len(samples.shape[:-1]))
+                if i not in self.seq_axis
+            ]
+        )
+        self.M_broadcast = mx.nd.broadcast_to(
+            M_expanded,
+            shape=_shape + list(proj_matrix_shape),
+        )
+
+        if self.seq_axis:
+            # bring the axis to iterate in the beginning
+            samples = mx.nd.moveaxis(
+                samples, self.seq_axis, list(range(len(self.seq_axis)))
+            )
+
+            out = []
+            for idx in product(
+                *[
+                    range(x)
+                    for x in [
+                        samples.shape[d] for d in range(len(self.seq_axis))
+                    ]
+                ]
+            ):
+                s = samples[idx]
+                out.append(
+                    mx.nd.linalg.gemm2(
+                        self.M_broadcast, s.expand_dims(-1)
+                    ).squeeze(axis=-1)
+                )
+
+            # put the axis in the correct order again
+            out = mx.nd.concat(*out, dim=0).reshape(samples.shape)
+            out = mx.nd.moveaxis(
+                out, list(range(len(self.seq_axis))), self.seq_axis
+            )
+            return out
+        else:
+            return mx.nd.linalg.gemm2(
+                self.M_broadcast, samples.expand_dims(-1)
+            ).squeeze(axis=-1)
+
+    def reconciliation_error(self, samples):
+        r"""
+        Computes the reconciliation error defined by the L-infinity norm of the constraint violation:
+                    || Ax ||_{\inf}
+
+        Parameters
+        ----------
+        samples
+            Samples
+
+        Returns
+        -------
+        Reconciliation error
+            Float
+
+        """
+        constraint_mat_shape = self.A.shape
+
+        A_expanded = self.A.expand_dims(axis=0)
+        A_broadcast = mx.nd.broadcast_to(
+            A_expanded, shape=samples.shape[0:1] + constraint_mat_shape
+        )
+        return mx.nd.max(
+            mx.nd.abs(
+                mx.nd.linalg_gemm2(A_broadcast, samples, transpose_b=True)
+            )
+        ).asnumpy()[0]
+
+    def loss(self, target, distr):
```

Comment:
Personally, I think it might be cleaner to split the different losses into sub-functions and just put the flag logic here. It makes it hard to read otherwise.

Reply:
yes, you are right.. This allows various combinations of computing loss (for ablation studies) and it is a bit involved. Refactored it as per your suggestion; check if it still needs to be improved.